File: go/aais/aais_test.go
package aais

import (
	"os"
	"testing"
	"time"
)

func fixture(t *testing.T, name string) Envelope {
	t.Helper()
	data, err := os.ReadFile("../../examples/" + name)
	if err != nil {
		t.Fatal(err)
	}
	envelope, err := Parse(data)
	if err != nil {
		t.Fatal(err)
	}
	return envelope
}

func TestSharedDigestAndValidation(t *testing.T) {
	request := fixture(t, "shell-approval.json")
	digest, err := ActionDigest(request.Request.Action)
	if err != nil || digest != request.Request.ActionDigest {
		t.Fatalf("digest mismatch: %s %v", digest, err)
	}
	_ = fixture(t, "approve-once.json")
}
func TestCanonicalKeyOrder(t *testing.T) {
	a, _ := ActionDigest(Action{Kind: "x", Name: "x", Summary: "x", Arguments: map[string]any{"b": 2, "a": 1}})
	b, _ := ActionDigest(Action{Kind: "x", Name: "x", Summary: "x", Arguments: map[string]any{"a": 1, "b": 2}})
	if a != b {
		t.Fatal("key order changed digest")
	}
}
func TestStoreApprovalReplayAndConflict(t *testing.T) {
	store := NewStore()
	request := fixture(t, "shell-approval.json")
	decision := fixture(t, "approve-once.json")
	if err := store.Add(request); err != nil {
		t.Fatal(err)
	}
	at := time.Date(2026, 8, 30, 18, 1, 0, 0, time.UTC)
	first, err := store.Decide(decision, at, nil)
	if err != nil || first.Resolution.Outcome != "approved" {
		t.Fatalf("%v %#v", err, first.Resolution)
	}
	second, err := store.Decide(decision, at, nil)
	if err != nil || second.Resolution.ID != first.Resolution.ID {
		t.Fatal("identical replay not idempotent")
	}
	changed := decision
	copy := *decision.Decision
	changed.Decision = &copy
	changed.Decision.ID = "dec_other"
	changed.Decision.Decision = "deny"
	if _, err := store.Decide(changed, at, nil); err == nil {
		t.Fatal("expected conflict")
	}
}
func TestExpiryAndUnofferedScope(t *testing.T) {
	request := fixture(t, "shell-approval.json")
	decision := fixture(t, "approve-once.json")
	expired := NewStore()
	_ = expired.Add(request)
	result, err := expired.Decide(decision, time.Date(2026, 8, 30, 19, 0, 0, 0, time.UTC), nil)
	if err != nil || result.Resolution.Outcome != "expired" {
		t.Fatal("expected expired")
	}
	store := NewStore()
	_ = store.Add(request)
	unoffered := decision
	copy := *decision.Decision
	unoffered.Decision = &copy
	unoffered.Decision.ID = "dec_persist"
	unoffered.Decision.Scope = "persistent"
	result, err = store.Decide(unoffered, time.Date(2026, 8, 30, 18, 1, 0, 0, time.UTC), nil)
	if err != nil || result.Resolution.Outcome != "invalid" {
		t.Fatalf("expected invalid: %v %#v", err, result.Resolution)
	}
}
func TestSnapshot(t *testing.T) {
	store := NewStore()
	_ = store.Add(fixture(t, "shell-approval.json"))
	at := time.Date(2026, 8, 30, 18, 1, 0, 0, time.UTC)
	snapshot := store.Snapshot("session_s1", at)
	if err := Validate(snapshot); err != nil || len(snapshot.Snapshot.Pending) != 1 {
		t.Fatalf("invalid snapshot: %v", err)
	}
	restored, err := FromSnapshot(snapshot)
	if err != nil || len(restored.Snapshot("", at).Snapshot.Pending) != 1 {
		t.Fatalf("snapshot restore failed: %v", err)
	}
}

func TestBuilders(t *testing.T) {
	at := time.Date(2026, 8, 30, 18, 0, 0, 0, time.UTC)
	expires := at.Add(10 * time.Minute)
	request, err := CreateRequest(CreateRequestOptions{Action: fixture(t, "shell-approval.json").Request.Action, Origin: Origin{Harness: "example", SessionID: "s-1"}, Risk: Risk{Level: "low", Reasons: []string{"test"}}, Choices: []Choice{{Decision: "approve", Scope: "once", Label: "Allow"}, {Decision: "deny", Scope: "once", Label: "Deny"}}, Sequence: 1, CreatedAt: at, ExpiresAt: &expires})
	if err != nil {
		t.Fatal(err)
	}
	_, err = CreateDecision(request, CreateDecisionOptions{Decision: "approve", Scope: "once", Actor: Actor{ID: "alex", Type: "human"}, Sequence: 2, DecidedAt: at.Add(time.Minute)})
	if err != nil {
		t.Fatal(err)
	}
}

File: java/src/test/java/io/github/alexmercedcoder/aais/AaisTest.java
package io.github.alexmercedcoder.aais;

import static org.junit.jupiter.api.Assertions.*;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.OffsetDateTime;
import org.junit.jupiter.api.Test;

class AaisTest {
  private static ObjectNode fixture(String name) throws IOException{return Aais.parse(Files.readString(Path.of("..","examples",name)));}
  @Test void validatesSharedDigest()throws IOException{ObjectNode request=fixture("shell-approval.json");assertEquals(request.path("request").path("action_digest").asText(),Aais.actionDigest(request.path("request").path("action")));fixture("approve-once.json");}
  @Test void canonicalizesKeyOrder()throws Exception{assertEquals(Aais.actionDigest(Aais.JSON.readTree("{\"b\":2,\"a\":1}")),Aais.actionDigest(Aais.JSON.readTree("{\"a\":1,\"b\":2}")));}
  @Test void approvesReplaysAndRejectsConflict()throws IOException{ApprovalStore store=new ApprovalStore();store.add(fixture("shell-approval.json"));ObjectNode decision=fixture("approve-once.json");OffsetDateTime at=OffsetDateTime.parse("2026-08-30T18:01:00Z");ObjectNode first=store.decide(decision,at,null);assertEquals("approved",first.path("resolution").path("outcome").asText());assertEquals(first,store.decide(decision,at,null));ObjectNode changed=decision.deepCopy();((ObjectNode)changed.path("decision")).put("id","dec_other").put("decision","deny");assertThrows(AaisException.class,()->store.decide(changed,at,null));}
  @Test void expiresAndRejectsUnofferedScope()throws IOException{ApprovalStore expired=new ApprovalStore();expired.add(fixture("shell-approval.json"));assertEquals("expired",expired.decide(fixture("approve-once.json"),OffsetDateTime.parse("2026-08-30T19:00:00Z"),null).path("resolution").path("outcome").asText());ApprovalStore store=new ApprovalStore();store.add(fixture("shell-approval.json"));ObjectNode decision=fixture("approve-once.json");((ObjectNode)decision.path("decision")).put("scope","persistent");assertEquals("invalid",store.decide(decision,OffsetDateTime.parse("2026-08-30T18:01:00Z"),null).path("resolution").path("outcome").asText());}
  @Test void snapshotIsValid()throws IOException{ApprovalStore store=new ApprovalStore();store.add(fixture("shell-approval.json"));OffsetDateTime at=OffsetDateTime.parse("2026-08-30T18:01:00Z");ObjectNode snapshot=store.snapshot("session_s1",at);assertEquals(1,snapshot.path("snapshot").path("pending").size());Aais.validate(snapshot);assertEquals(1,ApprovalStore.fromSnapshot(snapshot).snapshot(null,at).path("snapshot").path("pending").size());}
  @Test void buildersAreValid()throws IOException{ObjectNode source=fixture("shell-approval.json");OffsetDateTime at=OffsetDateTime.parse("2026-08-30T18:00:00Z");ObjectNode request=Aais.createRequest((ObjectNode)source.path("request").path("action"),(ObjectNode)source.path("request").path("origin"),(ObjectNode)source.path("request").path("risk"),(com.fasterxml.jackson.databind.node.ArrayNode)source.path("request").path("choices"),1,"session",at,at.plusMinutes(10));ObjectNode actor=Aais.JSON.createObjectNode().put("id","alex").put("type","human");Aais.validate(Aais.createDecision(request,"approve","once",actor,2,at.plusMinutes(1),null));}
}
